Ifu dev 20241115 1.13dev

VeeraRajasekhar opened the pull request:
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

ptrendx committed:
Changed VERSION to 1.13.0.dev

Signed-off-by: Przemyslaw Tredak <ptredak@nvidia.com>

yaox12 committed:
[PyTorch] Fix wgrads for GroupedLinear when weights don't require grad (#1258)

Fix wgrad for GroupedLinear when weights doesn't require grad

Signed-off-by: Xin Yao <xiny@nvidia.com>
Co-authored-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

yaox12 committed:
[Bugfix] Fix bias for 0-dim tensors in gemm (#1246)

* fix bias for 0-dim tensor

Signed-off-by: Xin Yao <xiny@nvidia.com>

* add check

Signed-off-by: Xin Yao <xiny@nvidia.com>

* use numel() instead of nullptr

Signed-off-by: Xin Yao <xiny@nvidia.com>

---------

Signed-off-by: Xin Yao <xiny@nvidia.com>

phu0ngng committed:
[TE/JAX] Enabling CudaGraph for custom calls with FFI  (#1228)

* register CmdBufferCompatible traits via C++ API

* renamed FFI_Traits

* use register_ffi_target()

---------

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

xrennvidia committed:
Fix seq_dim in CP implementation (#1264)

fix seq_dim in CP implementation

Signed-off-by: Xiaowei Ren <xren@nvidia.com>

timmoon10 committed:
[PyTorch] Reorganize L1 tests (#1255)

* Reorganize PyTorch L1 tests

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Move ONNX tests to L1

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Move FA version test to L3

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Limit parallel build jobs in FA version test

Signed-off-by: Tim Moon <tmoon@nvidia.com>

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>

timmoon10 committed:
[Paddle] Debug wheel test (#1265)

* Debug wheel test for PaddlePaddle

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix typo

Signed-off-by: Tim Moon <tmoon@nvidia.com>

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>

timmoon10 committed:
[PyTorch] Remove PyTorch L0 distributed test (#1273)

Remove PyTorch L0 distributed test

Forgot to remove in #1255.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

cyanguwa committed:
[PyTorch] Reduce the number of FA versions in L3 tests (#1280)

remove one FA version in the L3 test

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

mgoldfarb-nvidia committed:
Fused Attention Support 64-bit Ragged Offsets for Large THD Tensors (#1230)

* Use 64-bit offsets for cuDNN 9.5+
* Align workspace tensors to 16B.
* Fix bug where std::accumulate overflowed on large tensor shapes.
* Only support 64-bit offsets on arbitrary sequence length fp16 backend.

Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>

zlsh80826 committed:
[JAX] Skip V100 encoder tests (#1262)

* Skip encoder tests on V100

* Fix mulitprocessing jax.distributed.init

* Remove XLA xla_gpu_deterministic_ops which causes segfault

---------

Signed-off-by: Reese Wang <rewang@nvidia.com>

zlsh80826 committed:
Add THD + GQA supports (#1260)

Add THD + GQA supports for cuDNN >= 9.6

Signed-off-by: Reese Wang <rewang@nvidia.com>

mgoldfarb-nvidia committed:
[JAX] Fix correctness of JAX fused attention with CP and improve numerics check in unit tests (#1282)

Fix correctness of JAX fused attention with CP.

Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>

huanghua1994 committed:
[JAX] XLA Custom Calls with FFI for FusedAttnFwd, Quantize, Transpose, ActLuFP8, LayerNormForwardFP8FFI, and LayerNormBackwardFFI (#1263)

* Add TransposeFFI, test passed

Signed-off-by: Hua Huang <huah@nvidia.com>

* Add ActLuFP8FFI; fix TransposeFFI

Signed-off-by: Hua Huang <huah@nvidia.com>

* Add QuantizeFFI

Signed-off-by: Hua Huang <huah@nvidia.com>

* Add FusedAttnForwardFFI and some unit tests

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Minor fix

Signed-off-by: Hua Huang <huah@nvidia.com>

* Add LayerNormForwardFP8FFI & LayerNormBackwardFFI

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Revise FusedAttnForwardFFI()

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Add FFI_CudaGraph_Traits

All tests passed, ready for merge

Signed-off-by: Hua Huang <huah@nvidia.com>

* Bug fix for FFI data type mismatch

Also add a safeguard on the entrance to FFI function

Signed-off-by: Hua Huang <huah@nvidia.com>

---------

Signed-off-by: Hua Huang <huah@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

timmoon10 committed:
[Paddle] Update type names for Paddle 3.0 (#1286)

Update class names for Paddle 3.0

Signed-off-by: Tim Moon <tmoon@nvidia.com>

pggPL committed:
[Pytorch] Check gradient in test numerics (#1229)

* update test numerics

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* update test numerics

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* update test numerics

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Update tests/pytorch/test_numerics.py

Co-authored-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Signed-off-by: Paweł Gadziński <62263673+pggPL@users.noreply.github.com>

* tests fix

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

* Not passing CI fixes

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Not passing CI fixes

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

* Fix key

Co-authored-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

* fixes

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>

---------

Signed-off-by: Pawel Gadzinski <pgadzinski@nvidia.com>
Signed-off-by: Paweł Gadziński <62263673+pggPL@users.noreply.github.com>
Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>
Co-authored-by: Pawel Gadzinski <pgadzinski@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Co-authored-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

phu0ngng committed:
[TE/JAX] Update required JAX version for FFI custom calls with cudaGraph (#1285)

Update jax version for ffi

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

timmoon10 committed:
Support building documentation in Python 3.12 (#1274)

* Update dependencies for building documentation

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Patch RTD theme 3.0.0 to include version selector in sidebar

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Debug printing version in sidebar

Signed-off-by: Tim Moon <tmoon@nvidia.com>

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

cyanguwa committed:
[C/PyTorch] Add THD MQA/GQA (#1266)

* add THD MQA/GQA

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix nvte_get_fused_attn_backend

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

cyanguwa committed:
[C/PyTorch] Add max_t support for THD (#1244)

* WIP: add max_t support for THD

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* WIP: save tensors for debug and point to new FE

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix stats in bwd

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix stats in fwd

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* add docstring for DPA

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* add docstring

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* WIP: first try on adding max_b and max_t

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Revert "[pre-commit.ci] auto fixes from pre-commit.com hooks"

This reverts commit c3d522e9f5aef3c8ddfec5bf6ff24c3db97bb059.

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* Revert "WIP: first try on adding max_b and max_t"

This reverts commit 3bc01ebaf2aa846fd16634e2d33b0d0f5803a076.

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* update docstring and fix max_seqlen logic for thd

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* revert two lines of change in docstring

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* WIP: add get_max_b/t

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix max_seqlen code and docstring

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* sucess: add max_b/max_t

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* remove debug code

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* change max_b/max_t buckets

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix b vs orig_b

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* fix b vs orig_b with 0 fill

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* update FE for T3HD/TH3D

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* add max_b to conversion kernels

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix lint

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix changes after last merge

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* add Jax support for max_t

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* update FE to 1.8.0-rc

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* update FE to 1.8.0

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* code review/formating fixes

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix Stats shape for <9.6

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* return nullptr for offset_stats when cudnn < 9.6

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* add more version control

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

---------

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

timmoon10 committed:
[PyTorch] Remove fast param getter from modules (#1291)

* Add fallback for fast param getter

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Remove fast param getter

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix linter warning

Signed-off-by: Tim Moon <tmoon@nvidia.com>

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>

denera committed:
[C/PyTorch] Userbuffers and comm+GEMM overlap algorithms refactored and moved to TE/common (#1067)

* moved userbuffers code to TE/common

Signed-off-by: Alp Dener <adener@nvidia.com>

* moved comm+GEMM overlap code to TE/common

Signed-off-by: Alp Dener <adener@nvidia.com>

* removed PyTorch depdency from comm+GEMM overlap in TE/common

Signed-off-by: Alp Dener <adener@nvidia.com>

* added TE/PyTorch wrappers for refactored comm+GEMM overlap code in TE/common

Signed-off-by: Alp Dener <adener@nvidia.com>

* updated TE/PyTorch Python API to match the refactored comm+GEMM overlap code

Signed-off-by: Alp Dener <adener@nvidia.com>

* updated unit tests to work with refactored comm+GEMM overlap code

Signed-off-by: Alp Dener <adener@nvidia.com>

* added a pylint exception to comm+GEMM overlap test runner

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fixing linting errors

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* added documentation for te.initialize_ub

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fixed compile errors when building with NVTE_UB_WITH_MPI=1

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fixed default bootstrap backend

Signed-off-by: Alp Dener <adener@nvidia.com>

* switched default bootstrap backend priority to MPI > Gloo > NCCL

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* updated bootstrap backend documentation

Signed-off-by: Alp Dener <adener@nvidia.com>

* close UB bootstrap socket to avoid interfering with CUDA Multicast shareable file handle send/recv

Signed-off-by: Alp Dener <adener@nvidia.com>

* added torch::Tensor wrappers for communication buffer and atomic counters so PyTorch can factor externally allocated memory into its garbage collection threshold

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* automated handling of world, local and node ranks/sizes within C++ CommOverlapHelper to simplify Python function signatures

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fixed incorrect read of environment variables

Signed-off-by: Alp Dener <adener@nvidia.com>

* corrected priority for _SOCKET_IFNAME environment variables in UB bootstrapping

Signed-off-by: Alp Dener <adener@nvidia.com>

* moved multicast support check to cuda_runtime.h and replaced cudaDeviceGetProp call with cached sm_count()

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* removed commented out old code and replaced external collective function type defines with aliases

Signed-off-by: Alp Dener <adener@nvidia.com>

* compile-time CUDA version guard for CUDA Driver Multicast attribute

Signed-off-by: Alp Dener <adener@nvidia.com>

* added compile-time CUDA version guards to Multicast code in Userbuffers

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* condensed UB docs, corrected const violations

Signed-off-by: Alp Dener <adener@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fixed autodoc rst for UB calls, added CUDA version guard on Multicast UB kernels

Signed-off-by: Alp Dener <adener@nvidia.com>

* fixed incorrect UB type reporting for P2P overlaps, comment reformatting

Signed-off-by: Alp Dener <adener@nvidia.com>

* add docstring to tex.ubuf_built_with_mpi()

Signed-off-by: Alp Dener <adener@nvidia.com>

---------

Signed-off-by: Alp Dener <adener@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

cyanguwa committed:
[PyTorch] Skip t3hd/th3d for MQA/GQA tests (#1293)

skip some t3hd/th3d tests for MQA/GQA

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

cyanguwa committed:
Add check for GPU availability in attention (#1287)

* check if GPU is available

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Charlene Yang <8636796+cyanguwa@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

xrennvidia committed:
Add missed arguments of apply_rotary_pos_emb in MHA (#1296)

* add missed arguments of apply_rotary_pos_emb in MHA

Signed-off-by: Xiaowei Ren <xren@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* remove an unnecessary f

Signed-off-by: Xiaowei Ren <xren@nvidia.com>

* add one more assert for cp_group len

Signed-off-by: Xiaowei Ren <xren@nvidia.com>

---------

Signed-off-by: Xiaowei Ren <xren@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

mgoldfarb-nvidia committed:
[JAX] Consolidate FFI and old descriptor implementation for fused attention. (#1295)

Consolidate FFI and old descriptor impleemntation for fused attention.

Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>
Co-authored-by: Phuong Nguyen <36155692+phu0ngng@users.noreply.github.com>

phu0ngng committed:
[TE/JAX] Custom call with FFI - lowering all attributes with bind all (#1289)

* lowering a dict of attrs

* improve err message with line and func info

* implement a product() for ffi dimensions

---------

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

phu0ngng committed:
[TE/JAX] Disable FusedAttn with FFI by default (#1298)

* disable fused attn with ffi

---------

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

kunlunl committed:
Support using fp16 master weights and fp16/fp8 optimizer states in FusedAdam (#1078)

* Add precision aware fused adam

Signed-off-by: kunlunl <kunlunl@nvidia.com>

* Minor changes based on review comments.

Co-authored-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Signed-off-by: Kunlun Li <94586211+kunlunl@users.noreply.github.com>

---------

Signed-off-by: kunlunl <kunlunl@nvidia.com>
Signed-off-by: Kunlun Li <94586211+kunlunl@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>

phu0ngng committed:
[JAX] Fix for Disable FusedAttn with FFI by default (#1304)

rm default value for NVTE_JAX_FUSED_ATTN_WITH_FFI

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

denera committed:
[PyTorch] Missing intra-domain ranks list when initializing Userbuffers with data parallelism (#1305)

added missing list of intra-domain ranks when num_domains > 1 in initialize_ub

Signed-off-by: Alp Dener <adener@nvidia.com>

yaox12 committed:
[PyTorch] Make FP8 MHA work with RoPE when CP is on (#1297)

* Let fp8 mha work with rope when cp is on

Signed-off-by: Xin Yao <xiny@nvidia.com>

* fix and update ut

Signed-off-by: Xin Yao <xiny@nvidia.com>

---------

Signed-off-by: Xin Yao <xiny@nvidia.com>

kocchop committed:
[JAX] Expose context parallel params to jax DPA api (#1292)

Exposed context parallel params to DPA api

Signed-off-by: Md Fahim Faysal Khan <mdfahimfaysa@nvidia.com>
Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>

---------

Signed-off-by: Md Fahim Faysal Khan <mdfahimfaysa@nvidia.com>
Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>
Co-authored-by: Michael Goldfarb <mgoldfarb@nvidia.com>

huanghua1994 committed:
[TE/JAX] XLA FFI calls for layer norm and RMS norm (#1290)

* Add LayerNormForwardFFI(); add FFI calls in Python

Signed-off-by: Hua Huang <huah@nvidia.com>

* Add FFI for RMS norm, all tests passed

Signed-off-by: Hua Huang <huah@nvidia.com>

* Simplify layer & RMS norm FFI calls

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Simplify tensor size calculations

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Hua Huang <huah@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Phuong Nguyen <36155692+phu0ngng@users.noreply.github.com>

timmoon10 committed:
[PyTorch] Debug CUDA graph support with operation-based API (#1117)

* Debug CUDA graph support with operation-based API

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Refactoring CUDA graph tests

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Review suggestions from @ptrendx

Return default recipe from FP8GlobalStateManager.get_fp8_recipe if needed. Expand error message when failing to load FP8 state after capturing CUDA graph.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Avoid unnecessary recursion when saving/loading FP8 state

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix circular import

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

timmoon10 committed:
[PyTorch] Debug checkpointing with operation-based API (#1063)

* Debug checkpointing with operation-based API

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Store checkpoint FP8 state on CPU

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix bug where linear op was saving params multiple times

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix linter warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

timmoon10 committed:
[PyTorch] Normalization ops (#1033)

* Add layer norm op

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add FP8 cast op

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add tests for linear and layernorm with FP8 output

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* RMSNorm op

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Fix linter warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Replace LayerNorm module with LayerNorm op

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Replace RMSNorm module with RMSNorm op

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add AMP support

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Do not save autograd context if grad mode is disabled

Debugging ONNX export tests.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Forward args in pre_forward func to base op class

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Update to use QuantizedTensor class

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Apply suggestions from code review

Co-authored-by: Przemyslaw Tredak <ptrendx@gmail.com>
Signed-off-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Review suggestions from @ptrendx

Rename "CastFloat8" op to "Quantize". Add more fine-grained control for SM margin. Add docs for legacy sequence_parallel kwarg.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Fix linter warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Use weight dtype as default compute dtype

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix linter warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Signed-off-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Przemyslaw Tredak <ptrendx@gmail.com>

timmoon10 committed:
[PyTorch] Userbuffers support in operation-based API (#1142)

* Add Userbuffers support for column TP linear layer

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add Userbuffers support for row TP linear layer

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Interpret linear+RS as row TP linear

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add Userbuffers support for FP8 row TP linear layer

Assumes FP8 RS, which is not a good assumption.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Debug bug with incorrect bias pointers in UB GEMM

Bias pointers are not properly offset for different data chunks. Also removed logic for FP8 RS.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add Userbuffers support for linear dgrad

Test passes with row TP, fails with col TP.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add Userbuffers support for linear wgrad

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add support for grad bias

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fused cast-transpose-dbias

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Support case where wgrad is optional

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Expand documentation

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Fix linter warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Use recently added convenience functions in Float8Tensor

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Respect autograd dtype

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Fix missing imports

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Respect PyT autocast dtype in bprop

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix linter warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Debug merge conflicts

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Signed-off-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

timmoon10 committed:
Update list of CI users (#1316)

Signed-off-by: Tim Moon <tmoon@nvidia.com>

zlsh80826 committed:
[JAX] Add back the xla deterministic flag (#1301)

Add back the xla deterministic flag

Signed-off-by: Reese Wang <rewang@nvidia.com>
Co-authored-by: Phuong Nguyen <36155692+phu0ngng@users.noreply.github.com>

huanghua1994 committed:
[TE/JAX] XLA FFI calls for three cast transpose functions (#1310)

* FFI for some transpose & activation functions

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Remove comments in transformer_engine/jax/csrc/extensions/activation.cpp

Co-authored-by: Phuong Nguyen <36155692+phu0ngng@users.noreply.github.com>
Signed-off-by: Hua Huang <huangh1994@outlook.com>

---------

Signed-off-by: Hua Huang <huah@nvidia.com>
Signed-off-by: Hua Huang <huangh1994@outlook.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Phuong Nguyen <36155692+phu0ngng@users.noreply.github.com>

phu0ngng committed:
[JAX] Added prepare phase for the FusedAttnForwardFFI (#1313)

* added prepare phase for the FusedAttnForwardFFI

* enabled FusedAttnForwardFFI by default

* moved prepare phase into pybind

---------

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

phu0ngng committed:
[C] Separating cudnn common utils from fused_attn (#1314)

* split cudnn utils from fused_attn/util
---------

Signed-off-by: Phuong Nguyen <phuonguyen@nvidia.com>

mingxu1067 committed:
[JAX] Support Ring Attention (Context Parallelism) (#1059)

* Implement ring attention primative for Jax.

Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>
Signed-off-by: Ming Huang <mingh@nvidia.com>

---------

Signed-off-by: Michael Goldfarb <mgoldfarb@nvidia.com>
Signed-off-by: Ming Huang <mingh@nvidia.com>
Co-authored-by: Michael Goldfarb <mgoldfarb@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

ksivaman committed:
Convert non-kernel cuda files to cpp (#1322)

* Fix file extensions

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

* fix build

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* upgrade paddle container for CI

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

---------

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

huanghua1994 committed:
[TE/JAX] XLA FFI calls for Softmax and FusedAttnBackward (#1319)

* FFI for all softmax functions

Signed-off-by: Hua Huang <huah@nvidia.com>

* FFI for FusedAttnBackward and Dequantize

FusedAttnBackward passed all testes in test_fused_attn.py.
Dequantize is not used currently; finish it for completeness.

Signed-off-by: Hua Huang <huah@nvidia.com>

* Fix FusedAttnBackward FFI pybind & simplify

Signed-off-by: Hua Huang <huah@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Revert changes to tests/jax/test_fused_attn.py

Signed-off-by: Hua Huang <huah@nvidia.com>

---------

Signed-off-by: Hua Huang <huah@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Phuong Nguyen <36155692+phu0ngng@users.noreply.github.com>

jennifgcrl committed:
Fix an int conversion error (#1325)

fix an int conversion error

Signed-off-by: Jennifer Zhou <jennifer@jezh.me>

timmoon10 committed:
[PyTorch] Fix ONNX export bug with operation-based API (#1320)

Debug ONNX export with te.Sequential

ONNX export assumes that all state dict objects are tensor, even extra state.

Signed-off-by: Tim Moon <tmoon@nvidia.com>

timmoon10 committed:
[PyTorch] Remove special handling for FP8 params in FP8 recipe infrastructure (#1326)

* Remove manual FP8 scale update for FP8 params

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* lint

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

ksivaman committed:
[PyTorch] Fix multiple calls to saved_tensors in CP attention (#1334)

* Limit to one call of ctx.saved_tensors per autograd bwd

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

timmoon10 committed:
[PyTorch] Activation operations (#1164)

* Add activation ops

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Fix lint warnings

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Fix linter warning

Signed-off-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>

* Update to use QuantizedTensor

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Respect PyTorch autograd dtype

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Rename CastFloat8 op to Quantize

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* Add support for fused dSwiGLU-cast-transpose

Signed-off-by: Tim Moon <tmoon@nvidia.com>

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Signed-off-by: Tim Moon <tmoon@nvidia.com>
Signed-off-by: Tim Moon <4406448+timmoon10@users.noreply.github.com>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

VeeraRajasekhar committed:
Merge commit '20b0473cd1a5e999d3f5996f5c45809410f45455' into IFU-dev-20241115-1.13dev
Resolved all the conflicts from the IFU

ipanfilo commented on examples/jax/encoder/common.py:
```diff
@@ -0,0 +1,14 @@
+@lru_cache
+def is_bf16_supported():
+    """Return if BF16 has hardware supported"""
+    gpu_arch = get_device_compute_capability(0)
```
This code needs hip specific path

VeeraRajasekhar replied:
https://github.com/ROCm/TransformerEngine/blob/06ed6c9c1c1af39ade705eb66a93e62e8675a9ee/transformer_engine/pytorch/utils.py#L257

Using this

VeeraRajasekhar replied:
Fixed

ipanfilo commented on setup.py:
```diff
@@ -60,6 +60,13 @@ def run(self):
 
 def setup_common_extension() -> CMakeExtension:
     """Setup CMake extension for common library"""
+    cmake_flags = ["-DCMAKE_CUDA_ARCHITECTURES={}".format(cuda_archs())]
```
How does it work on ROCm?

VeeraRajasekhar replied:
Resolved this in latest update commit **Fixed setup.py conflict with DCMAKE_CUDA_ARCHITECTURES**

VeeraRajasekhar committed:
TE IFU 1.13 integration changes

parent e29f7ca048c2f0e54a90bf10465ce7e2c2077cc0
author Veera Rajasekhar <veerarajasekharreddy.gopu@amd.com> 1741200865 +0000
committer Veera Rajasekhar <veerarajasekharreddy.gopu@amd.com> 1743434534 -0500

Fixed setup.py conflict with DCMAKE_CUDA_ARCHITECTURES

Added rocm path checks for jax bf16

Updated renamed test path in ci/pytorch.sh script

Fixed rocm specific info for custom_call.py

Fixed formatting common/CmakeLists.txt

removed cudnn_utils.cpp from rocm CMake

Added HIP guard for cuda_runtime.cpp file

updated rocm guards for gemm files, updated pybinder files

Updated hip code in jax/cpp_extensions/attention.py

Remove targets gfx940 and gfx941. (#136)

Fix hipblaslt algo save regression from commit ed158e1d8

[ROCm] integrate ck varlen feature with THD layout into TE (#135)

* [ROCm] integrate ck varlen feature with THD layout into TE

* ROCm: address the reviwer comments

* [ROCm] address more reviewer comments

Enable Flash and Fused attn with CP and THD (#142)

Fixed rocm path variables with proper guards, Guarded gemm file. Able to build TE for pytorch

Fixed jax build issue.

Updated the rmsnorm.py with triton based kernels

Added CommOverlapAlgo dummy placeholder

[Cherry-pick] cherry pick "[MoE][PyTorch] Add mask-based MoE permutation" (#140)

* [Cherry-pick] cherry pick "[MoE][PyTorch] Add mask-based MoE permutation" from NV upstream

* resolve reviewer issues and fix unit tests

* resolve reviewer issues

---------

Co-authored-by: RuibinZhang <Ruibin.Zhang@amd.com>

Fixed build w/o AOTriton support

[ROCm] hotfix for THD in TE CK: avoid reading total_seqlen_q and total_seqlen_kv directly on host (#146)

* [ROCm] hotfix for THD in TE CK: avoid reading total_seqlen_q and total_seqlen_kv directly on host

* [ROCm] address reviewer comments

* [ROCm] resolve more comments

Jax build fix

Jax pytests fp8 fix

torch adam fix for fp8 range on MI300.

Fix for ring attention tests in test_distributed_fused_attn.py

Fixed fp8 fnuz in pytorch utils.py

Added rocm guards for rmsnorm fp8 support

Remove undesired default rng_state update: fix numerical test on MI308

Updated Copyrights

Jax allgather unbalanced workaround fix

Fix JAX 0.4.35 JIT w/a on MI308

[ROCm] make ck float to bf16 conversion as a compile-time option (#150)

* [ROCm] make ck float to bf16 conversion as a compile-time option

* [ROCm] remove the default definition in ck_fused_attn cmakefile

TE core: add support for ROCm FP8 FNUZ and OCP formats

Disable calling CK v3 bwd on SBHD formats (#151)

* [ROCm] disable ck v3 bwd on SBHD format

* [ROCm] add logging for sbhd v3 disabling and fix some nvte ck logging

Enable pad_between_seqs features in ROCm TE CK backend (#147)

* [ROCm] enable pad between seq

* [ROCm] move pad_between_seqs implementation to transformer_engine/common/fused_attn_rocm

* [ROCm] improve the padding/unpadding kernel and skip test praxis layer with CK backend

* [ROCm] optimize lse thd to [b, h, s] conversion

* [ROCm] address Ilya's reviewer comments

* [ROCm] move softmax lse converstion to te common

* [ROCm] remove unnecessaries headers

* [ROCm] warp->wavefront, lane->workitem

* [ROCm] address Wen's reviewer comments

* [ROCm] more comments for remove/add padding

Workaround test hang with JAX 0.4.35. Add log for test failure (#158)

Git ignore hipify results and ROCm binaries

Jax test_fused_attn tests fix

ipanfilo commented on tests/pytorch/distributed/test_fusible_ops.py:
It requires ci/pytorch.sh update

VeeraRajasekhar replied:
Updated renamed test path in ci/pytorch.sh script

ipanfilo commented on tests/pytorch/utils.py:
```diff
@@ -0,0 +1,85 @@
+        return dict(rtol=1.3e-6, atol=1e-5)
+    if dtype == torch.float64:
+        return dict(rtol=1e-7, atol=1e-7)
+    if dtype == torch.float8_e4m3fn:
```
tolerances for float8_e4m3fnuz and e5m2fnuz are required

VeeraRajasekhar replied:
https://github.com/ROCm/TransformerEngine/blob/7678277d1c11ab9d3b1008309c1359acd42e25b6/tests/pytorch/triton_kernels/test_rmsnorm_triton.py#L41

ipanfilo commented on transformer_engine/common/CMakeLists.txt:
```diff
@@ -148,10 +149,16 @@ if(USE_CUDA)
        fused_rope/fused_rope.cu
-       recipe/delayed_scaling.cu)
+       recipe/delayed_scaling.cu
+     comm_gemm_overlap/userbuffers/ipcsocket.cc
```
Is it really formatted this way?

VeeraRajasekhar replied:
Fixed formatting common/CmakeLists.txt

ipanfilo commented on transformer_engine/common/CMakeLists.txt:
```diff
@@ -148,10 +149,16 @@ if(USE_CUDA)
 else()
   list(APPEND transformer_engine_SOURCES
+       # Changed
+       cudnn_utils.cpp
```
cudnn_utils.cpp does not seem buildable on ROCm

VeeraRajasekhar replied:
Yes, removed from the list

ipanfilo commented on transformer_engine/common/util/cuda_runtime.cpp:
```diff
@@ -102,6 +103,31 @@ int sm_count(int device_id) {
 }
 
+bool supports_multicast(int device_id) {
+#if CUDART_VERSION >= 12010
```
Should HIP_PLATFORM_AMD guard be used here or whole method moved under guard below?

VeeraRajasekhar replied:
Added HIP guards for the new additions of multi-cast functions.

ipanfilo commented on transformer_engine/common/util/pybind_helper.h:
```diff
@@ -0,0 +1,79 @@
+      .value("NVTE_THD_T2HD", NVTE_QKV_Layout::NVTE_THD_T2HD)                                 \
+      .value("NVTE_THD_TH2D", NVTE_QKV_Layout::NVTE_THD_TH2D)                                 \
+      .value("NVTE_THD_THD_THD", NVTE_QKV_Layout::NVTE_THD_THD_THD);                          \
+  pybind11::enum_<NVTE_Fused_Attn_Backend>(m, "NVTE_Fused_Attn_Backend")                      \
```
ROCm specific values are different here

VeeraRajasekhar replied:
Updated with rocm specific values.

ipanfilo commented on transformer_engine/jax/cpp_extensions/attention.py:
```diff
@@ -274,21 +279,23 @@ def abstract(
             config.window_size,
         ).get_fused_attn_backend()
 
-        if not is_hip_extension():
```
Incorrect merge removed HIP code

VeeraRajasekhar replied:
Updated the Hip code caused by merging

ipanfilo commented on transformer_engine/jax/cpp_extensions/custom_call.py:
Incorrect merge removed ROCm code

VeeraRajasekhar replied:
Fixed rocm specific info for custom_call.py

ipanfilo commented on transformer_engine/pytorch/csrc/common.h:
```diff
@@ -28,6 +28,7 @@
 #include <transformer_engine/activation.h>
 #include <transformer_engine/cast.h>
 #include <transformer_engine/cast_transpose_noop.h>
+#include <transformer_engine/comm_gemm_overlap.h>
```
Non ROCm path only

VeeraRajasekhar replied:
Updated with rocm guard

ipanfilo commented on transformer_engine/pytorch/csrc/extensions/pybind.cpp:
```diff
@@ -342,48 +258,61 @@ PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
-      .value("NVTE_CAUSAL_BOTTOM_RIGHT_MASK", NVTE_Mask_Type::NVTE_CAUSAL_BOTTOM_RIGHT_MASK)
-      .value("NVTE_PADDING_CAUSAL_BOTTOM_RIGHT_MASK",
-             NVTE_Mask_Type::NVTE_PADDING_CAUSAL_BOTTOM_RIGHT_MASK);
+  py::class_<CommOverlapHelper>(m, "CommOverlapHelper")
```
CommOverlap is not supported on ROCm

VeeraRajasekhar replied:
Updated with the rocm guard

ipanfilo commented on transformer_engine/pytorch/module/rmsnorm.py:
```diff
@@ -5,233 +5,159 @@
-        inputmat = cast_if_needed(inputmat, activation_dtype)
-        rmsnorm_weight = cast_if_needed(rmsnorm_weight, activation_dtype)
-
-        use_rmsnorm_triton = bool( int(os.environ.get('NVTE_USE_RMSNORM_TRITON', '0')) )
```
ROCm specific code is lost on merging

VeeraRajasekhar replied:
Fixed and updated with rocm specific code

ipanfilo commented on transformer_engine/pytorch/optimizers/fused_adam.py:
```diff
@@ -134,14 +176,208 @@ def __init__(
+            torch.float16: torch.full(
+                [1], torch.finfo(torch.float16).max / 2.0, dtype=torch.float32
+            ),
+            torch.uint8: torch.full([1], 448.0, dtype=torch.float32),
```
MI300 FP8 has max 240.0

VeeraRajasekhar replied:
Fixed this.

ipanfilo commented on ci/jax.sh:
```diff
@@ -76,8 +76,14 @@ run_test_config_mgpu() {
-    run 3 test_distributed_fused_attn.py
+    run 3 test_distributed_fused_attn.py -k 'not test_context_parallel_ring_attn'
+    # Test ring attention with and without scan loop
+    NVTE_FUSED_RING_ATTENTION_USE_SCAN=0 run 3 test_distributed_fused_attn.py -k test_context_parallel_ring_attn
```
Is NVTE_FUSED_RING_ATTENTION_USE_SCAN used in any other tests except test_context_parallel_ring_attn ? And what is default value for it?

VeeraRajasekhar replied:
The default value is 1. No, only the test_context_parallel_ring_attn uses this

ipanfilo commented on examples/jax/encoder/common.py:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Copyright

ipanfilo commented on status.txt:
```diff
@@ -0,0 +1,32 @@
+Search for #Changed to find the changes you made.
```
Why is this file part of PR?

VeeraRajasekhar replied:
Sorry, removed this. Added for my notes.

ipanfilo commented on setup.py:
```diff
@@ -60,7 +60,6 @@ def run(self):
 
 def setup_common_extension() -> CMakeExtension:
     """Setup CMake extension for common library"""
-    cmake_flags = ["-DCMAKE_CUDA_ARCHITECTURES={}".format(cuda_archs())]
```
Potentially CUDA build is broken

VeeraRajasekhar replied:
This was there in the else part below, Don't know why this is showing as removed

ipanfilo commented on tests/jax/test_custom_call_compute.py:
```diff
@@ -797,8 +798,8 @@ def test_transpose(self, in_dtype, input_shape, transpose_axis):
-            pytest.param(jnp.float8_e4m3fn, id="output_float8_e4m3fn"),
-            pytest.param(jnp.float8_e5m2, id="output_float8_e5m2"),
+            pytest.param(jnp_float8_e4m3_type, id="output_float8_e4m3fn"),
+            pytest.param(jnp_float8_e4m3_type, id="output_float8_e5m2"),
```
typo: should be e5m2

ipanfilo commented on tests/pytorch/fused_attn/test_fused_attn.py:
```diff
@@ -674,6 +674,8 @@ def test_dpa_qkv_layout_thd(dtype, model_configs, model, qkv_layout, pad_between
     if config.num_heads != config.num_gqa_groups and "3" in qkv_layout:
         pytest.skip("qkv_layout not applicable for MQA/GQA")
     pad_between_seqs = True
+    if (pad_between_seqs==False and get_cudnn_version() < (9, 3, 0)):
```
Why is it returned? Wasn't it removed in upstream?

VeeraRajasekhar replied:
yes, I have updated this.

ipanfilo commented on transformer_engine/common/CMakeLists.txt:
```diff
@@ -150,15 +149,14 @@ if(USE_CUDA)
   list(APPEND transformer_engine_SOURCES
-       # Changed
-       cudnn_utils.cpp
+      #  cudnn_utils.cpp
```
Not needed here

VeeraRajasekhar replied:
Removed this

ipanfilo commented on transformer_engine/common/util/cuda_runtime.cpp:
```diff
@@ -127,6 +130,8 @@ bool supports_multicast(int device_id) {
 #endif
 }
+#endif // __HIP_PLATFORM_AMD__
+
```
It can be merged with the next #ifndef

VeeraRajasekhar replied:
Merged this.

ipanfilo commented on transformer_engine/common/util/cuda_runtime.cpp:
```diff
@@ -14,7 +14,9 @@
 #include "../common.h"
 #include "../util/cuda_driver.h"
 #include "../util/system.h"
+#ifndef __HIP_PLATFORM_AMD__
```
Not needed. The header declares methods defined below

VeeraRajasekhar replied:
Okay, removed this.

ipanfilo commented on transformer_engine/common/util/cuda_runtime.h:
Copyright year

ipanfilo commented on transformer_engine/common/util/cuda_runtime.cpp:
Copyright year

ipanfilo commented on transformer_engine/common/util/pybind_helper.h:
Copyright

ipanfilo commented on transformer_engine/jax/cpp_extensions/custom_call.py:
Copyright

ipanfilo commented on transformer_engine/jax/csrc/extensions.h:
Copyright year

ipanfilo commented on transformer_engine/jax/csrc/extensions/cudnn.cpp:
Copyright and it looks like the whole file can be guarded

ipanfilo commented on transformer_engine/jax/csrc/extensions/pybind.cpp:
Copyright

ipanfilo commented on transformer_engine/jax/util.py:
Copyright

ipanfilo commented on transformer_engine/pytorch/csrc/common.h:
Copyright

ipanfilo commented on transformer_engine/pytorch/csrc/extensions/comm_gemm_overlap.cpp:
Copyright

ipanfilo reviewed (commented):
All modified files need updated copyright

ipanfilo commented on examples/jax/encoder/common.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Please follow the same pattern for copyright as in other files: 
`# This file was modified for portability to AMDGPU`
`# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.`
`# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.`

ipanfilo commented on tests/pytorch/fused_attn/test_fused_attn.py:
```diff
@@ -670,11 +670,16 @@ def test_dpa_qkv_layout(dtype, model_configs, model, qkv_layout):
 @pytest.mark.parametrize("pad_between_seqs", [False, True])
 def test_dpa_qkv_layout_thd(dtype, model_configs, model, qkv_layout, pad_between_seqs):
     """Test DotProductAttention module with different QKV layouts"""
-    if (pad_between_seqs==False and get_cudnn_version() < (9, 3, 0)):
```
This method modification breaks (partially reverts) 06ed6c9c1c1af39ade705eb66a93e62e8675a9ee .

VeeraRajasekhar replied:
fixed it, Reverted the changes.

ipanfilo commented on tests/pytorch/utils.py:
```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright pattern

ipanfilo commented on transformer_engine/common/util/pybind_helper.h:
```diff
@@ -0,0 +1,102 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright format

ipanfilo commented on transformer_engine/jax/csrc/extensions/cudnn.cpp:
```diff
@@ -0,0 +1,25 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright format

ipanfilo commented on transformer_engine/pytorch/csrc/extensions.h:
Copyright year

ipanfilo commented on transformer_engine/pytorch/csrc/extensions.h:
```diff
@@ -510,5 +518,186 @@ void multi_tensor_sgd_cuda(int chunk_size, at::Tensor noop_flag,
+                        size_t workspaceSize, bool accumulate, bool use_split_accumulator,
+                        at::Tensor rs_output);
+};  // CommOverlapP2P
+#endif
```
Better add comment `#endif // !USE_ROCM`

ipanfilo commented on transformer_engine/pytorch/csrc/extensions/comm_gemm_overlap.cpp:
```diff
@@ -0,0 +1,482 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright format

ipanfilo commented on transformer_engine/pytorch/csrc/extensions/comm_gemm_overlap.cpp:
```diff
@@ -0,0 +1,482 @@
+  int output_c_dim1 = _ubuf.size(1);
+  return torch::from_blob(ubuf_wt_ptr, {output_c_dim0, output_c_dim1}, _ubuf_torch.options());
+}
+#endif
```
Better add comment `#endif // !USE_ROCM`

ipanfilo commented on transformer_engine/pytorch/csrc/extensions/pybind.cpp:
Copyright year

ipanfilo commented on transformer_engine/pytorch/ops/basic/rmsnorm.py:
```diff
@@ -0,0 +1,305 @@
+# Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright pattern

ipanfilo commented on transformer_engine/pytorch/optimizers/fused_adam.py:
```diff
@@ -1,13 +1,19 @@
 # Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright format

ipanfilo commented on ci/jax.sh:
```diff
@@ -74,10 +74,30 @@ run_test_config() {
+    
+    _ver=$(pip show jaxlib | grep Version)
+    case "$_ver" in
+    *0.4.35*)
```
The w/a seems not working on MI308

VeeraRajasekhar replied:
fixed by iliya

ipanfilo commented on ci/jax.sh:
```diff
@@ -74,10 +74,30 @@ run_test_config() {
+        ;;
+    *0.4.31*)
+        #Workaround for JAX 0.4.31 regression: crash in test_destributed_fused_attn and test_distributed_layernorm_mlp
+        #TODO: Will remove this after fixing jax 0.4.35 test issue with UNBALANCED tests in test_distributed_fused_attn::test_context_parallel_allgather_attn
```
Why is this comment here?

ipanfilo commented on tests/pytorch/fused_attn/test_fused_attn.py:
```diff
@@ -670,8 +670,11 @@ def test_dpa_qkv_layout(dtype, model_configs, model, qkv_layout):
+        pytest.skip("qkv_layout not applicable for MQA/GQA")
     if (pad_between_seqs==False and get_cudnn_version() < (9, 3, 0)):
-        pytest.skip("cuDNN 9.3.0+ is required to run pad_between_seqs = False");
+         pytest.skip("cuDNN 9.3.0+ is required to run pad_between_seqs = False");
```
Incorrect align

wenchenvincent reviewed (changes requested):
Overall it looks good. I have one more suggestion before merging. There are 30+ commits that you made from early March to now in the IFU beside those commits from the upstream. Can you squash those commits into one using interactive rebase? That would make the commit history much cleaner. Otherwise those commits would be interleaved with the upstream commits in our future IFUs.

VeeraRajasekhar committed:
Merge branch 'dev' into IFU-dev-20241115-1.13dev
